### PR TITLE
Allow host spill store to fit a buffer larger than configured max size

### LINKED ISSUE #784: [FEA] Allow Host Spilling to be more dynamic
**Is your feature request related to a problem? Please describe.**
Recently new code was added to branch-0.3 to allow spilling from the GPU to the CPU for things like `GpuCoalesceBatch`  The issue is that the default amount of memory allocated on the host for spilling is only 1GB and it is common to be able to have a buffer that is larger than that.  This means that even though we could spill to host memory we cannot for any buffer larger than the current allocated buffer size.

**Describe the solution you'd like**
If the buffer is larger than we can accommodate it would be nice if the `RapidsHostMemoryStore` would allocate a temporary buffer to hold the data and then instantly start to spill it to disk. Ideally for flow control we would allow only one of these to be outstanding at a time.

**Describe alternatives you've considered**
We could leave it as is, but then users will likely have to reconfigure their job by default to have it run correctly instead of just crashing with a different memory error.

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #784.

This updates the host memory spill store to do two things:
- It no longer fails if a buffer is spilled that is larger than the configured maximum size.  Instead it will spill everything it can first then allocates a host buffer to hold the incoming device buffer.  It logs a warning when this occurs for the first time.
- It proactively spills buffers before allocating a new buffer in an attempt to keep under the configured maximum size.  Previously it would only spill if the pinned memory pool could not allocate the buffer *and* the internal pool of host memory could not allocate the buffer, meaning it could end up using significantly more total host memory for spilled buffers than was configured.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala [hunk 1]
```diff
@@ -31,15 +31,27 @@ class RapidsHostMemoryStore(
...
-    require(size <= maxSize, s"allocating a buffer of $size bytes beyond max of $maxSize")
+    // spill to keep within the targeted size
+    synchronousSpill(math.max(maxSize - size, 0))
+
     var buffer: HostMemoryBuffer = null
     while (buffer == null) {
       buffer = PinnedMemoryPool.tryAllocate(size)
       if (buffer != null) {
-        return (buffer, true)
+        return (buffer, false)
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:45]: seems like this should return `true` here? It got a buffer from the pool if I read it correctly.

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:45]: Not the internal pool maintained by the host spill store but the external pinned memory pool.  Essentially this flag is for tracking if it needs to release it with the address allocator when the buffer is freed.

**Comment by jlowe** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:45]: ok got it.. the rest get freed via ref count

**Review by abellina — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala (hunk 1 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:45]: The perhaps we need some better comments in allocateHostBuffer to explain this.

**Review by revans2 — APPROVED** [review]

**Pull request merged**